lib/mwUtil.js

```diff
@@ -311,7 +311,8 @@ mwUtil.createRelativeTitleRedirect = (path, req, options) => {
             + new URI(pathPatternAfterTitle, newReqParams, true).toString().substr(1)
             + mwUtil.getQueryString(req);
     } else {
-        location = backString + encodeURIComponent(newReqParams.title) + mwUtil.getQueryString(req);
+        location = backString + encodeURIComponent(newReqParams.title).replace(/%23/g, '#')
```

Comment:
According to RFC the fragment can not contain the `#` and it must be percent-encoded. So we only need to replace the first occurrence of `%23`, cause if we replace all something like `https://bla.com/lala%23bla%23bla` will become `https://bla.com/lala#bla#bla` which is icorrect, it should be https://bla.com/lala#bla%23bla

Reply:
Really good catch! {{done}}

Reply:
As it turns out, MW does a bad job at it - https://en.wikipedia.beta.wmflabs.org/w/index.php?title=User:Pchelolo/Redirect_Test_Hash&redirect=no :)